- In this study, we are going to make Exploratory Data Analysis (EDA) with the Amazon.com's bestseller books
- Study aims to be beginner friendly and give as much as possible explanation for each step on the way.
- Study's dataset has 550 books along with their ratings, price, publication year, authors' name and genre.
- Data includes 2009-2019 best seller books.

- Let's import the required libraries

In [1]:
import pandas as pd
import numpy as np


import plotly 
import plotly.express as px
import plotly.graph_objs as go
import plotly.offline as py
from plotly.offline import iplot
from plotly.subplots import make_subplots
import plotly.figure_factory as ff

### Overview Stage

- Read the csv
- Look for basic information about the dataset

In [2]:
df = pd.read_csv('../input/amazon-top-50-bestselling-books-2009-2019/bestsellers with categories.csv')
df.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

In [3]:
df.shape

(550, 7)

- So we have 550 book and 7 features to work on

In [4]:
df.isnull().sum()

Name           0
Author         0
User Rating    0
Reviews        0
Price          0
Year           0
Genre          0
dtype: int64

- We have a very clean dataset, which is very rare in the real world. 
- So enjoy working data without any missing value in it.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         550 non-null    object 
 1   Author       550 non-null    object 
 2   User Rating  550 non-null    float64
 3   Reviews      550 non-null    int64  
 4   Price        550 non-null    int64  
 5   Year         550 non-null    int64  
 6   Genre        550 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 30.2+ KB


- We have 4 numeric variable
- Also we 3 non-numeric variable.
- As a data type, everything seems quite OK.

In [6]:
df.describe()

User Rating       Reviews       Price         Year
count   550.000000    550.000000  550.000000   550.000000
mean      4.618364  11953.281818   13.100000  2014.000000
std       0.226980  11731.132017   10.842262     3.165156
min       3.300000     37.000000    0.000000  2009.000000
25%       4.500000   4058.000000    7.000000  2011.000000
50%       4.700000   8580.000000   11.000000  2014.000000
75%       4.800000  17253.250000   16.000000  2017.000000
max       4.900000  87841.000000  105.000000  2019.000000

Before going further, let's summarize what we have got from the dataset.

- Our dataset has 550 books from different authors and genres.

- Object data type variable (genre) can be grouped and see the differences among them.

- Reviews and price columns most probably have outliers. (Mean- Median difference, difference between 75% and maximum value, difference between %25 and minimum value)

-  Numerical variables deserves special attention for further analysis.


- Everything seems OK.  Let's move on to the next step: **analysis part**.

### Analysis Part

#### **Author**

In [7]:
df['Author'].nunique()

248

- 248 differnt authors are in the dataset.

#### **Genre**

In [8]:
df['Genre'].value_counts(normalize=True)

Non Fiction    0.563636
Fiction        0.436364
Name: Genre, dtype: float64

- I was expecting more genre, it is quite surprising.
- Anyway, Genre is still good to use to see differences between two category.

In [9]:
fig = px.histogram(df, x="Genre", title='Genre')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

#### **User Rating**

In [10]:
df['User Rating'].describe()

count    550.000000
mean       4.618364
std        0.226980
min        3.300000
25%        4.500000
50%        4.700000
75%        4.800000
max        4.900000
Name: User Rating, dtype: float64

- Mean and median score are quite close to each other. (Median_ 4.7, Mean=4.618)
- Since median score is bigger than mean score we can expect outlier from the minimum side. 
- Most probably we will have left skewed distribution. 
- But still we can expect close to normal distribution of the variable.
- Let's see it

In [11]:
fig = px.histogram(df, x= 'User Rating', title='User Rating', marginal="box", hover_data = df[['Name','Author']])
fig.show()

- Yeap, as expected we have several outliers from the minimum side.
- Slightly left skewed distribution, but still close to the normal distribution
- Oh wait...
- Oh no!! J.K.Rowling's 'Casual Vacancy' got the lowest rating.

#### **Reviews**

In [12]:
df['Reviews'].describe()

count      550.000000
mean     11953.281818
std      11731.132017
min         37.000000
25%       4058.000000
50%       8580.000000
75%      17253.250000
max      87841.000000
Name: Reviews, dtype: float64

- We have huge difference between mean and median values (mean = 11953, median=8580)
- It has highly skewed distribution with the outliers on the maximum side.
- We can expect highly right skewed distribution with possible outliers in the maximum side.
- Let' see it.

In [13]:
fig = px.histogram(df, x= 'Reviews', title='Reviews', marginal="box", hover_data = df[['Name','Author']])
fig.show()

- As expected, highly right skewed distribution with the outliers on the maximum side.

- By the way, I saw that Kristin Hannah's 'The Nightingale' is among the outliers with 49K plus reviews. It is a beautiful novel to read.

#### **Price**

In [14]:
df['Price'].describe()

count    550.000000
mean      13.100000
std       10.842262
min        0.000000
25%        7.000000
50%       11.000000
75%       16.000000
max      105.000000
Name: Price, dtype: float64

- We can expect slightly rightly skewed distribution
- Still, distribution will be close to the normal distribution.
- We can expect outliers on the maximum side.
- Yeah, we have also free books. 
- Let me note them to check it after the analys and if any of them still free, I would be happy to have it.

In [15]:
fig = px.histogram(df, x= 'Price', title='Price', marginal="box", hover_data = df[['Name','Author']])
fig.show()

- As we expected, we have slightly right skewed distribution with outliers on the maximum side.
- Also we have 13 counts of books on the $0-1 range. 13 free books, sounds good to me.

- Befor moving on the details, let's see the correlation matrix for our dataset

In [16]:
df.drop('Year', axis=1).corr()

User Rating   Reviews     Price
User Rating     1.000000 -0.001729 -0.133086
Reviews        -0.001729  1.000000 -0.109182
Price          -0.133086 -0.109182  1.000000

In [17]:
index_vals = df['Genre'].astype('category').cat.codes

fig = go.Figure(data=go.Splom(
                dimensions=[dict(label='User Rating',
                                 values=df['User Rating']),
                            dict(label='Reviews',
                                 values=df['Reviews']),
                            dict(label='Price',
                                 values=df['Price'])],
                showupperhalf=False, 
                text=df['Name'],
                marker=dict(color=index_vals,
                            showscale=False, # colors encode categorical variables
                            line_color='white', line_width=0.5)
                ))


fig.update_layout(
    title='Books',
    width=1000,
    height=1000,
)

fig.show()

- There isn't any significant correlation to consider for further analysis.

- After getting overall picture about the data, we can go into more details.

In [18]:
genre_by_year = df.groupby('Year')['Genre'].value_counts().reset_index(level=0).rename(columns={'Genre': 'Genre count'}, index={'index': 'Genre'})
genre_by_year

Year  Genre count
Genre                         
Non Fiction  2009           26
Fiction      2009           24
Non Fiction  2010           30
Fiction      2010           20
Non Fiction  2011           29
Fiction      2011           21
Non Fiction  2012           29
Fiction      2012           21
Non Fiction  2013           26
Fiction      2013           24
Fiction      2014           29
Non Fiction  2014           21
Non Fiction  2015           33
Fiction      2015           17
Non Fiction  2016           31
Fiction      2016           19
Non Fiction  2017           26
Fiction      2017           24
Non Fiction  2018           29
Fiction      2018           21
Non Fiction  2019           30
Fiction      2019           20

#### **Movie Genre in Each Year**

In [19]:
fig = px.line(genre_by_year, x='Year', y='Genre count', color= genre_by_year.index, title='Movies By Genre in Each Year')
fig.show()

- Non fiction books  are sharply descreased on the 2014, and then sharply increased on 2015.
- Fiction books significantly increased on 2014 and then sharply decreased on 2015.
- Both fiction and non-fiction books have inconsistency on their counts by year.

#### **Price of Books in Each Year**

In [20]:
fig = px.scatter(df, x='Year', y='Price', title='Price of the Books in Each Year', hover_data = df[['Name','Author']])
fig.show()

- Prices of the books are quite on the same range by year with several outliers.
- We have 2013 and 2014 books at the price of $105 from American Psychiatric Association.

#### **Number of Reviews in Each Year**

In [21]:
fig = px.scatter(df, x='Year', y='Reviews', title='Number of Reviews in Each Year', color='Genre',hover_data = df[['Name','Author']])
fig.show()

- Quite same distribution by each year, especialy after 2010.
- Several outliers affect the distribution, as we have mentioned before.

#### **User Rating in Each Year**

In [22]:
fig = px.scatter(df, x='Year', y='User Rating', title='User Rating in Each Year',color='Genre', hover_data = df[['Name','Author']])
fig.show()

- User rating has almost same distribution on each year with a quite few outliers.

### **Top 20 Higly Rated Books**

In [23]:
top_20 = df.sort_values('User Rating', ascending=False)[:20]
top_20

Name                 Author  \
431                                 The Magnolia Story            Chip Gaines   
87   Dog Man: Lord of the Fleas: From the Creator o...             Dav Pilkey   
85   Dog Man: Fetch-22: From the Creator of Captain...             Dav Pilkey   
84   Dog Man: Brawl of the Wild: From the Creator o...             Dav Pilkey   
83   Dog Man: Brawl of the Wild: From the Creator o...             Dav Pilkey   
82   Dog Man: A Tale of Two Kitties: From the Creat...             Dav Pilkey   
81   Dog Man and Cat Kid: From the Creator of Capta...             Dav Pilkey   
252                          Oh, the Places You'll Go!              Dr. Seuss   
476                        The Very Hungry Caterpillar             Eric Carle   
477                        The Very Hungry Caterpillar             Eric Carle   
478                        The Very Hungry Caterpillar             Eric Carle   
479                        The Very Hungry Caterpillar             Eric Carle   
480                        The Very Hungry Caterpillar             Eric Carle   
481                        The Very Hungry Caterpillar             Eric Carle   
482                        The Very Hungry Caterpillar             Eric Carle   
486                   The Wonderful Things You Will Be  Emily Winfield Martin   
487                   The Wonderful Things You Will Be  Emily Winfield Martin   
488                   The Wonderful Things You Will Be  Emily Winfield Martin   
489                   The Wonderful Things You Will Be  Emily Winfield Martin   
86   Dog Man: For Whom the Ball Rolls: From the Cre...             Dav Pilkey   

     User Rating  Reviews  Price  Year        Genre  
431          4.9     7861      5  2016  Non Fiction  
87           4.9     5470      6  2018      Fiction  
85           4.9    12619      8  2019      Fiction  
84           4.9     7235      4  2019      Fiction  
83           4.9     7235      4  2018      Fiction  
82           4.9     4786      8  2017      Fiction  
81           4.9     5062      6  2018      Fiction  
252          4.9    21834      8  2019      Fiction  
476          4.9    19546      5  2013      Fiction  
477          4.9    19546      5  2014      Fiction  
478          4.9    19546      5  2015      Fiction  
479          4.9    19546      5  2016      Fiction  
480          4.9    19546      5  2017      Fiction  
481          4.9    19546      5  2018      Fiction  
482          4.9    19546      5  2019      Fiction  
486          4.9     8842     10  2016      Fiction  
487          4.9     8842     10  2017      Fiction  
488          4.9     8842     10  2018      Fiction  
489          4.9     8842     10  2019      Fiction  
86           4.9     9089      8  2019      Fiction

In [24]:
fig = px.bar(top_20, x='Name', y= 'User Rating',  hover_data = top_20[['Year','Genre', 'Price']], color='Genre')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- Several books with their yearly editions got the most reviews from the readers.
- Only one book from non-fiction genre find place in the top 20 highly rated book list.
- Also maximum price for the books in the top 20 list is $10

#### **Lowest Rated 20 Books**

In [25]:
bottom_20 = df.sort_values('User Rating')[:20]
bottom_20

Name          Author  \
353                                 The Casual Vacancy    J.K. Rowling   
132                         Go Set a Watchman: A Novel      Harper Lee   
106  Fifty Shades of Grey: Book One of the Fifty Sh...       E L James   
107  Fifty Shades of Grey: Book One of the Fifty Sh...       E L James   
393  The Goldfinch: A Novel (Pulitzer Prize for Fic...     Donna Tartt   
22                                           Allegiant   Veronica Roth   
392  The Goldfinch: A Novel (Pulitzer Prize for Fic...     Donna Tartt   
364                       The Elegance of the Hedgehog  Muriel Barbery   
137                                          Gone Girl   Gillian Flynn   
136                                          Gone Girl   Gillian Flynn   
304                                StrengthsFinder 2.0          Gallup   
135                                          Gone Girl   Gillian Flynn   
306                                StrengthsFinder 2.0          Gallup   
307                                StrengthsFinder 2.0          Gallup   
311                                StrengthsFinder 2.0          Gallup   
312                                StrengthsFinder 2.0          Gallup   
308                                StrengthsFinder 2.0          Gallup   
305                                StrengthsFinder 2.0          Gallup   
154  Harry Potter and the Cursed Child, Parts 1 & 2...    J.K. Rowling   
309                                StrengthsFinder 2.0          Gallup   

     User Rating  Reviews  Price  Year        Genre  
353          3.3     9372     12  2012      Fiction  
132          3.6    14982     19  2015      Fiction  
106          3.8    47265     14  2012      Fiction  
107          3.8    47265     14  2013      Fiction  
393          3.9    33844     20  2014      Fiction  
22           3.9     6310     13  2013      Fiction  
392          3.9    33844     20  2013      Fiction  
364          4.0     1859     11  2009      Fiction  
137          4.0    57271      9  2014      Fiction  
136          4.0    57271     10  2013      Fiction  
304          4.0     5069     17  2009  Non Fiction  
135          4.0    57271     10  2012      Fiction  
306          4.0     5069     17  2011  Non Fiction  
307          4.0     5069     17  2012  Non Fiction  
311          4.0     5069     17  2016  Non Fiction  
312          4.0     5069     17  2017  Non Fiction  
308          4.0     5069     17  2013  Non Fiction  
305          4.0     5069     17  2010  Non Fiction  
154          4.0    23973     12  2016      Fiction  
309          4.0     5069     17  2014  Non Fiction

In [26]:
fig = px.bar(bottom_20, x='Name', y= 'User Rating',  hover_data = bottom_20[['Year','Genre', 'Price']], color='Genre')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- Lowest rating is 3.3.
- Only 1 book from non-fiction genre in the list.
- Maximum price of the book in the list is $20

#### **Top 20 Reviewed Books**

In [27]:
top_20_reviews = df.sort_values('Reviews', ascending=False)[:20]
top_20_reviews

Name          Author  \
534                            Where the Crawdads Sing     Delia Owens   
382                              The Girl on the Train   Paula Hawkins   
383                              The Girl on the Train   Paula Hawkins   
32                                            Becoming  Michelle Obama   
33                                            Becoming  Michelle Obama   
137                                          Gone Girl   Gillian Flynn   
135                                          Gone Girl   Gillian Flynn   
136                                          Gone Girl   Gillian Flynn   
368                             The Fault in Our Stars      John Green   
367                             The Fault in Our Stars      John Green   
365                             The Fault in Our Stars      John Green   
366                             The Fault in Our Stars      John Green   
437                           The Nightingale: A Novel  Kristin Hannah   
438                           The Nightingale: A Novel  Kristin Hannah   
106  Fifty Shades of Grey: Book One of the Fifty Sh...       E L James   
107  Fifty Shades of Grey: Book One of the Fifty Sh...       E L James   
433                                        The Martian       Andy Weir   
20                         All the Light We Cannot See   Anthony Doerr   
21                         All the Light We Cannot See   Anthony Doerr   
338                                      The Alchemist    Paulo Coelho   

     User Rating  Reviews  Price  Year        Genre  
534          4.8    87841     15  2019      Fiction  
382          4.1    79446     18  2015      Fiction  
383          4.1    79446      7  2016      Fiction  
32           4.8    61133     11  2018  Non Fiction  
33           4.8    61133     11  2019  Non Fiction  
137          4.0    57271      9  2014      Fiction  
135          4.0    57271     10  2012      Fiction  
136          4.0    57271     10  2013      Fiction  
368          4.7    50482     13  2014      Fiction  
367          4.7    50482      7  2014      Fiction  
365          4.7    50482     13  2012      Fiction  
366          4.7    50482     13  2013      Fiction  
437          4.8    49288     11  2015      Fiction  
438          4.8    49288     11  2016      Fiction  
106          3.8    47265     14  2012      Fiction  
107          3.8    47265     14  2013      Fiction  
433          4.7    39459      9  2015      Fiction  
20           4.6    36348     14  2014      Fiction  
21           4.6    36348     14  2015      Fiction  
338          4.7    35799     39  2014      Fiction

In [28]:
fig = px.bar(top_20_reviews, x='Name', y= 'Reviews',  hover_data = top_20_reviews[['Year','Genre', 'Price']], color='Genre')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- Several books with their different editions got the most reviews.
- Only one book from non-fiction genre find place in the list.
- Maximum price of the book is one of the my favorite book, 'The Alchemist' by $35.

#### **Lowest Number Reviewed 20 Books**

In [29]:
bottom_20_reviews = df.sort_values('Reviews')[:20]
bottom_20_reviews

Name             Author  \
78   Divine Soul Mind Body Healing and Transmission...       Zhi Gang Sha   
300  Soul Healing Miracles: Ancient and New Sacred ...       Zhi Gang Sha   
121                    George Washington's Sacred Fire  Peter A. Lillback   
512                             True Compass: A Memoir  Edward M. Kennedy   
359  The Daily Show with Jon Stewart Presents Earth...        Jon Stewart   
11   A Patriot's History of the United States: From...   Larry Schweikart   
39   Broke: The Plan to Restore Our Trust, Truth an...         Glenn Beck   
27                 Autobiography of Mark Twain, Vol. 1         Mark Twain   
264  Percy Jackson and the Olympians Paperback Boxe...       Rick Riordan   
31   Barefoot Contessa, How Easy Is That?: Fabulous...         Ina Garten   
316  Teach Like a Champion: 49 Techniques that Put ...         Doug Lemov   
317  Teach Like a Champion: 49 Techniques that Put ...         Doug Lemov   
91   Eat This Not That! Supermarket Survival Guide:...    David Zinczenko   
25   Arguing with Idiots: How to Stop Small Minds a...         Glenn Beck   
374  The Five Love Languages: How to Express Heartf...       Gary Chapman   
444  The Official SAT Study Guide, 2016 Edition (Of...  The College Board   
346  The Book of Basketball: The NBA According to T...       Bill Simmons   
92   Eat This, Not That! Thousands of Simple Food S...    David Zinczenko   
299                                  Sookie Stackhouse   Charlaine Harris   
193                               JOURNEY TO THE ICE P          RH Disney   

     User Rating  Reviews  Price  Year        Genre  
78           4.6       37      6  2009  Non Fiction  
300          4.6      220     17  2013  Non Fiction  
121          4.5      408     20  2010  Non Fiction  
512          4.5      438     15  2009  Non Fiction  
359          4.4      440     11  2010  Non Fiction  
11           4.6      460      2  2010  Non Fiction  
39           4.5      471      8  2010  Non Fiction  
27           4.2      491     14  2010  Non Fiction  
264          4.8      548      2  2010      Fiction  
31           4.7      615     21  2010  Non Fiction  
316          4.4      637     20  2010  Non Fiction  
317          4.4      637     20  2011  Non Fiction  
91           4.5      720      1  2009  Non Fiction  
25           4.6      798      5  2009  Non Fiction  
374          4.6      803      9  2009  Non Fiction  
444          4.3      807     36  2016  Non Fiction  
346          4.7      858     53  2009  Non Fiction  
92           4.3      956     14  2009  Non Fiction  
299          4.7      973     25  2009      Fiction  
193          4.6      978      0  2014      Fiction

In [30]:
fig = px.bar(bottom_20_reviews, x='Name', y= 'Reviews',  hover_data = bottom_20_reviews[['Year','Genre', 'Price']], color='Genre')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- Zhi Gang Sha's two differnt books got the lowest reviews.
- Only 3 fiction books are in the lowest reviewed book list.
- Other books come from non-fiction genre.
- We can make an assumption about it. But still we need other data to support our assumptions.

- OK Let's see the top 25 Authors in our dataset

#### **Top 25 Authors** 

In [31]:
top_25_authors = df['Author'].value_counts()[:25]
top_25_authors

Jeff Kinney                           12
Rick Riordan                          11
Suzanne Collins                       11
Gary Chapman                          11
American Psychological Association    10
Gallup                                 9
Dr. Seuss                              9
Rob Elliott                            8
Dav Pilkey                             7
Stephenie Meyer                        7
Stephen R. Covey                       7
Eric Carle                             7
Bill O'Reilly                          7
Harper Lee                             6
E L James                              6
The College Board                      6
Don Miguel Ruiz                        6
Stieg Larsson                          6
J.K. Rowling                           6
Sarah Young                            6
John Grisham                           5
R. J. Palacio                          5
Laura Hillenbrand                      5
John Green                             5
Dale Carnegie   

In [32]:
fig = px.bar(top_25_authors, x= top_25_authors.index, y=top_25_authors.values, title='Top 25 Authors',labels={'y':'Number of Books', 'index':'Author'})
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- Let's see these top 25 authors, user ratings and reviews scores.

In [33]:
top_25_authors_ratings = df[df['Author'].isin(top_25_authors.index)][['Author','User Rating', 'Reviews']]
top_25_authors_ratings_grouped=top_25_authors_ratings.groupby('Author')[['User Rating','Reviews']].mean().sort_values('Reviews', ascending=False)

In [34]:
fig = px.bar(top_25_authors_ratings_grouped, x= top_25_authors_ratings_grouped.index, y='User Rating', title='Top 25 Authors with USer Rating')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- As expected, they have very high user rating (4.9 to 4)

In [35]:
top_25_authors_reviews_grouped=top_25_authors_ratings.groupby('Author')[['User Rating', 'Reviews']].mean().sort_values('Reviews', ascending=False)
fig = px.bar(top_25_authors_ratings_grouped, x= top_25_authors_ratings_grouped.index, y='Reviews', title='Top 25 Authors with Reviews')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

- Thanks for the dataset contibutor for this data. I really enjoyed working on it.

- It was a quite pleasure to share with you this detailed, beginner friendly EDA. Thanks for your time.

- All the best 